Support for ConcurrentMap

**Is your feature request related to a problem? Please describe.**
The OpenLiberty website was recently updated to use JSON-B over JSON-P to return data from a REST API. The JSON-B object in question has a field of type ConcurrentMap. This was correctly serialized. As a parallel activity I'm writing some integration tests and for the first time tried to deserialize the JSON-B object. This failed with the following error:

    [ERROR] io.openliberty.website.it.OpenLibertyEndpointIT.testGetReleases  Time elapsed: 0.201 s  <<< ERROR!
    javax.ws.rs.client.ResponseProcessingException: Problem with reading the data, class io.openliberty.website.data.BuildData, ContentType: application/json.
	at io.openliberty.website.it.OpenLibertyEndpointIT.testGetReleases(OpenLibertyEndpointIT.java:41)
    Caused by: javax.json.bind.JsonbException: Internal error: Can not set java.util.concurrent.ConcurrentMap field io.openliberty.website.data.BuildData.builds to java.util.HashMap
	at io.openliberty.website.it.OpenLibertyEndpointIT.testGetReleases(OpenLibertyEndpointIT.java:41)

It seems a JSON-B object with a ConcurrentMap can be serialized, but cannot be deserialized. It should be possible to do both.

**Describe the solution you'd like**
Support for JSON-B fields to contain a ConcurrentMap

**Describe alternatives you've considered**
In this case I will convert to Map to get it to work. This would not work as well if I were doing concurrent updates to the map though as I'd have to use syncronization around the map.

--- a/src/main/java/org/eclipse/yasson/internal/serializer/MapDeserializer.java
+++ b/src/main/java/org/eclipse/yasson/internal/serializer/MapDeserializer.java
@@ -16,8 +16,12 @@ import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.NavigableMap;
 import java.util.SortedMap;
 import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentSkipListMap;
 
 import jakarta.json.bind.serializer.JsonbDeserializer;
 import jakarta.json.stream.JsonParser;
@@ -48,7 +52,6 @@ public class MapDeserializer<T extends Map<?, ?>> extends AbstractContainerDeser
      *
      * @param builder {@link DeserializerBuilder} used to build this instance
      */
-    @SuppressWarnings({"unchecked", "rawtypes"})
     protected MapDeserializer(DeserializerBuilder builder) {
         super(builder);
         mapValueRuntimeType = getRuntimeType() instanceof ParameterizedType
@@ -67,6 +70,13 @@ public class MapDeserializer<T extends Map<?, ?>> extends AbstractContainerDeser
     }
 
     private Map getMapImpl(Class ifcType, DeserializerBuilder builder) {
+        if (ConcurrentMap.class.isAssignableFrom(ifcType)) {
+            if (SortedMap.class.isAssignableFrom(ifcType) || NavigableMap.class.isAssignableFrom(ifcType)) {
+                return new ConcurrentSkipListMap<>();
+            } else {
+                return new ConcurrentHashMap<>();
+            }
+        }
         // SortedMap, NavigableMap
         if (SortedMap.class.isAssignableFrom(ifcType)) {
             Class<?> defaultMapImplType = builder.getJsonbContext().getConfigProperties().getDefaultMapImplType();

--- a/src/test/java/org/eclipse/yasson/defaultmapping/collections/CollectionsTest.java
+++ b/src/test/java/org/eclipse/yasson/defaultmapping/collections/CollectionsTest.java
@@ -12,15 +12,36 @@
 
 package org.eclipse.yasson.defaultmapping.collections;
 
-import org.junit.jupiter.api.*;
-import static org.junit.jupiter.api.Assertions.*;
 import static org.eclipse.yasson.Jsonbs.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.math.BigDecimal;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.EnumMap;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
 
 import org.eclipse.yasson.TestTypeToken;
 import org.eclipse.yasson.defaultmapping.generics.model.Circle;
-
-import java.math.BigDecimal;
-import java.util.*;
+import org.junit.jupiter.api.Test;
 
 /**
  * Default mapping arrays/collections/enums tests.
@@ -254,5 +275,51 @@ public class CollectionsTest {
         assertEquals(TreeMap.class, result.getClass());
         assertEquals("abc", result.get("first"));
         assertEquals("def", result.get("second"));
+    }
+    
+    public static class ConcurrentMapContainer {
+    	public ConcurrentMap<String, String> concurrentMap;
+    	public ConcurrentHashMap<String,String> concurrentHashMap;
+    	public ConcurrentNavigableMap<String, String> concurrentNavigableMap;
+    	public ConcurrentSkipListMap<String, String> concurrentSkipListMap;
+    }
+    
+    @Test
+    public void testConcurrentMaps() {
+    	// ConcurrentMap
+    	ConcurrentMapContainer c = new ConcurrentMapContainer();
+    	c.concurrentMap = new ConcurrentHashMap<String, String>();
+    	c.concurrentMap.put("foo", "fooVal");
+    	c.concurrentMap.put("bar", "barVal");
+    	String expectedJson = "{\"concurrentMap\":{\"bar\":\"barVal\",\"foo\":\"fooVal\"}}";
+    	assertEquals(expectedJson, defaultJsonb.toJson(c));
+    	assertEquals(c.concurrentMap, defaultJsonb.fromJson(expectedJson, ConcurrentMapContainer.class).concurrentMap);
+    	
+    	// ConcurrentHashMap
+    	c = new ConcurrentMapContainer();
+    	c.concurrentHashMap = new ConcurrentHashMap<String, String>();
+    	c.concurrentHashMap.put("foo", "fooVal2");
+    	c.concurrentHashMap.put("bar", "barVal2");
+    	expectedJson = "{\"concurrentHashMap\":{\"bar\":\"barVal2\",\"foo\":\"fooVal2\"}}";
+    	assertEquals(expectedJson, defaultJsonb.toJson(c));
+    	assertEquals(c.concurrentHashMap, defaultJsonb.fromJson(expectedJson, ConcurrentMapContainer.class).concurrentHashMap);
+    	
+    	// ConcurrentNavigableMap
+    	c = new ConcurrentMapContainer();
+    	c.concurrentNavigableMap = new ConcurrentSkipListMap<String, String>();
+    	c.concurrentNavigableMap.put("foo", "fooVal3");
+    	c.concurrentNavigableMap.put("bar", "barVal3");
+    	expectedJson = "{\"concurrentNavigableMap\":{\"bar\":\"barVal3\",\"foo\":\"fooVal3\"}}";
+    	assertEquals(expectedJson, defaultJsonb.toJson(c));
+    	assertEquals(c.concurrentNavigableMap, defaultJsonb.fromJson(expectedJson, ConcurrentMapContainer.class).concurrentNavigableMap);
+    	
+    	// ConcurrentSkipListMap
+    	c = new ConcurrentMapContainer();
+    	c.concurrentSkipListMap = new ConcurrentSkipListMap<String, String>();
+    	c.concurrentSkipListMap.put("foo", "fooVal4");
+    	c.concurrentSkipListMap.put("bar", "barVal4");
+    	expectedJson = "{\"concurrentSkipListMap\":{\"bar\":\"barVal4\",\"foo\":\"fooVal4\"}}";
+    	assertEquals(expectedJson, defaultJsonb.toJson(c));
+    	assertEquals(c.concurrentSkipListMap, defaultJsonb.fromJson(expectedJson, ConcurrentMapContainer.class).concurrentSkipListMap);
     }
 }

